### PR TITLE
Add batch optimiser for symbolic expressions

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

File: tests/symengine/batch_optimizer_01_1a.cc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// Test that all of the low-level math operations function as expected
+// and that their values and derivatives can be computed using the
+// optimizer.
+// For this test the optimizer is in mode "off", i.e. plain old
```
Comment (Rombur) [inline tests/symengine/batch_optimizer_01_1a.cc:20]: There is no `off` mode anymore

File: tests/symengine/batch_optimizer_01_1b.cc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// Test that all of the low-level math operations function as expected
+// and that their values and derivatives can be computed using the
+// optimizer.
+// For this test the optimizer is in mode "off", i.e. plain old
```
Comment (Rombur) [inline tests/symengine/batch_optimizer_01_1b.cc:20]: There is no more `off` mode anymore.

File: tests/symengine/batch_optimizer_01_1a.cc (hunk 1 above)
Comment (jppelteret) [reply tests/symengine/batch_optimizer_01_1a.cc]: Thanks, I've fixed this up.

**Comment by jppelteret** [comment]
> I trust you that it's the same thing I've already reviewed

Indeed, it is. I just squashed the fixes back into the main commits.

> You need to add an entry in the changelog.

Thanks for the reminder! I've now added one.

**Comment by Rombur** [comment]
I'll merge this PR tomorrow unless someone as an objection (you have 24 hours to review 25 thousands line so don't forget the Red Bull)

**Comment by jppelteret** [comment]
Alternatively, the main class documentation and the doxygen module are
relatively light reading :-D

**Comment by bangerth** [comment]
Thanks, @jppelteret and @Rombur for working through this all!

I think we really ought to make writing a tutorial for all of this a priority. step-33 just isn't very good in this regard. Maybe an adaptation of step-15 might work?

**Comment by jppelteret** [comment]
> I think we really ought to make writing a tutorial for all of this a priority.

The plan was to write an introductory tutorial at the hackathon. Let's see if I can still manage that. There's also the [code gallery example for AD](https://github.com/dealii/code-gallery/tree/master/Quasi_static_Finite_strain_Compressible_Elasticity), and in particular [its extension](https://github.com/dealii/code-gallery/pull/50). I think I'll just touch and an merge that PR, and at least we can reference it for now.

**Pull request merged**